ktor-client/ktor-client-features/ktor-client-auth/common/src/io/ktor/client/features/auth/providers/BasicAuthProvider.kt

```diff
@@ -69,13 +118,19 @@ public class BasicAuthProvider(
     }
 
     override suspend fun addRequestHeaders(request: HttpRequestBuilder) {
-        request.headers[HttpHeaders.Authorization] = constructBasicAuthValue()
+        val credentials = tokensHolder.loadToken() ?: return
+        request.headers[HttpHeaders.Authorization] = constructBasicAuthValue(credentials)
     }
 
-    internal fun constructBasicAuthValue(): String {
-        val authString = "$username:$password"
-        val authBuf = authString.toByteArray(defaultCharset).encodeBase64()
-
-        return "Basic $authBuf"
+    override suspend fun refreshToken(response: HttpResponse): Boolean {
+        tokensHolder.setToken(credentials)
+        return true
     }
 }
+
+internal fun constructBasicAuthValue(credentials: BasicAuthCredentials): String {
+    val authString = "${credentials.username}:${credentials.password}"
```

Comment:
Could you tell me why we move it outside of the class?

Reply:
It was easier to test it this way. This function doesn't need reference to an instance.